backend/queues/txProcessor.js

```diff
@@ -0,0 +1,135 @@
+ */
+function attachToQueue() {
+  const queue = queues['txQueue']
+  queue.process(processor)
+  queue.resume() // Start if paused
+}
+
+/**
+ * Waits for a blockchain transaction to get confirmed,
+ * then enqueues a job to make an offer.
+ *
+ * @param {Object} job: Bull job object.
+ * job.data is expected to have the following fields:
+ *   {number} shopId: Id of the shop.
+ *   {string} txHash: hash of the tx to watch.
+ *   {string} fromAddress: Address sending the transaction.
+ *   {string} toAddress: Address receiving the transaction.
+ *   {string} encryptedDataIpfsHash: IPFS hash of the encrypted data.
+ * @returns {Promise<TransactionReceipt||null>}
+ * @throws
+ */
+async function processor(job) {
+  const queueLog = (progress, str) => {
+    job.log(str)
+    job.progress(progress)
+  }
+  const jobId = `${job.queue.name}-${job.id}` // Prefix with queue name since job ids are not unique across queues.
+
+  const { shopId, txHash, fromAddress, encryptedDataIpfsHash } = job.data
+  log.info(`txProcessor for job with data: ${job.data}`)
+
+  // TODO: check the validity of the IPFS data?
+
+  // Load the associated shop.
+  const shop = await Shop.findOne({ where: { id: shopId } })
+  if (!shop) {
+    throw new Error(`Failed loading shop with id ${shopId}`)
+  }
+
+  // Load the active network and create an associated ether.js provider.
+  const network = await Network.findOne({
+    where: { networkId: shop.networkId, active: true }
+  })
+  if (!network) {
+    throw new Error(`Failed loading active network with id ${shop.networkId}`)
+  }
+  const provider = new ethers.providers.JsonRpcProvider(network.provider)
+
+  // Load the transaction from the DB.
+  const transaction = await Transaction.findOne({
+    networkId: network.id,
+    shopId,
+    hash: txHash
+  })
+  if (!transaction) {
+    throw new Error(
+      `No transaction found in the DB with hash ${txHash} for shop ${shopId} and wallet ${fromAddress}`
+    )
+  }
+
+  // Check the transaction is not already marked as Confirmed.
+  // It would indicate the job is being retried despite the logic having
+  // successfully run. Possibly a corner case due to Bull's at-least-once guarantee.
+  if (transaction.status === TransactionStatuses.Confirmed) {
+    log.error(`Transaction ${transaction.id} already confirmed.`)
+    queueLog(100, `Transaction already confirmed`)
+    return null
+  }
+
+  // Load the tx from the blockchain based on its hash.
+  queueLog(25, 'Loading tx from the blockchain')
+  const tx = await provider.getTransaction(txHash)
+  if (!tx) {
+    throw new Error(`Transaction with hash ${txHash} not found`)
+  }
+  log.info(`Loaded tx with hash ${txHash} from the network`)
+
+  // TODO: check to and from address.
+
+  // Wait for the tx to get mined.
+  queueLog(50, `Waiting for tx ${txHash} to get confirmed`)
+  log.info(`Waiting for tx ${txHash} confirmation...`)
+  const receipt = await tx.wait(NUM_BLOCKS_CONFIRMATION)
```

Comment:
Wondering if this is a good idea for this queue.  It locks up a worker and may timeout and error anyway(unsure).  Perhaps this thing would work better with a high amount of retries and a tuned backoff?

Reply:
Yes good point. I'm not too worried about locking up a worker given our current minuscule volume.
But if we scale up that could become a bottleneck. I have added a comment int the code to document this.